[sosreport] Concurrently run plugins

TurboTurtle opened the pull request:
Adds concurrency to sos, defaulting to using 4 threads to run plugins. Adds a `--threads` option to control this value. Also adds a `timeout` value to plugins (defaulting to 300 second) so if the plugin takes longer than the timeout to execute the thread/plugin is terminated, though we'll still get the output collected up until that point.

Output from sosreport while running plugins now looks like this:

``` 
Setting up archive ...
 Setting up plugins ...
 Running plugins. Please wait ...

  Starting 1/89 : abrt            [Running: abrt]
  Starting 2/89 : anaconda        [Running: abrt anaconda]
  Starting 3/89 : anacron         [Running: abrt anaconda anacron]
  Starting 4/89 : apache          [Running: abrt anaconda anacron apache]
  Starting 5/89 : auditd          [Running: abrt anaconda apache auditd]
  Starting 6/89 : autofs          [Running: abrt apache auditd autofs]
[...]
  Starting 89/89: xfs             [Running: logs systemd virsh xfs]
  Finishing plugins               [Running: logs systemd virsh]
  Finishing plugins               [Running: logs virsh]
  Finishing plugins               [Running: logs]
  Finished running plugins

```

and if a plugin times out, this is printed

```
 Plugin logs timed out

```

Note: there can be some weirdness with writing to the archive if a timeout is set extremely low. I don't imagine this ever being the case in reality (by low I'm talking a few seconds),  but thought it worth mentioning.

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

pmoravec commented:
Side-effect from a user experience: pressing Ctrl+C does not further work.

If I realize I fired wrong `sosreport` / with wrong commands or on wrong system, there is no way to stop it via Ctrl+C (until I am very lucky and fast and press it before spawning the threads, I guess).

TurboTurtle commented:
Yeah, that's a limitation of using a multiprocessing pool (or dummy pool). This is resolved with the `futures` package in python-3.x, but for 2.7 it's kinda the state of things. You can still pause and kill the process, which i realize is not ideal, but IMO is a reasonable trade-off for the performance gain.

battlemidget commented:
What is the performance gain of this?

pmoravec commented:
Performance gain much depends on the use case. I.e. I let it run on 2 systems:

- a smaller with foreman (where foreman plugin takes its time): improvement from 2m14.626s to 1m11.951s of real time
- a bigger one with foreman and docker (that had some docker cmds timeouting due to an unrelated issue): improvement from 8m51.320s to 5m20.962s of real time

Having many plugins taking similar execution time, the gain can be even considerably bigger than above, having one leeengthy plugin, no (big) gain.

TurboTurtle commented:
Let me clarify one thing actually: the loss of Ctrl+C is I believe _only_ apparent due to the wrapping via `_collect_plugin()` so that we can implement a plugin-wide timeout. If we drop that, then we should be able to use a regular `multiprocessing.Pool()` rather than a dummy pool which would in turn allow us to initialize the workers to ignore Ctrl+C and pass the signal to the parent ( `collect()`) to be handled in the main thread. 

Since the dummy pool uses a separate thread itself, we can't use `signal.signal()` for the workers in the Pool to ignore Ctrl+C (since `signal` can only be used in the main thread.


That being said, I am personally of the opinion that implementing a plugin-wide timeout (and thus being able to avoid a large number of sos hangs due to hung plugin commands) is a benefit that outweighs the loss of Ctrl+C and having to SIGKILL an unwanted sosreport process (or just riding it out).

TurboTurtle commented:
After discussing this a bit, I've updated the PR to use the `concurrent.futures` module. Now, instead of using a `multiprocessing.Pool`, the patch uses a `ThreadPoolExecutor` which allows us to successfully capture keyboard interrupts in the main thread and exit appropriately.

Two things to note here:

1. This will add a dep on `python-futures` for any python2 environment. As of python3, `futures` is in the standard library. This dependency is also why the python2.7 Travis job is failing.

2. For the keyboard interrupt to work, we need to use `os._exit()` instead of the `_exit()` method already defined in `SoSReport` since we need to call `shutdown(wait=True)` in order to close the executor, and we cannot override that later with an immediate shutdown.

TurboTurtle commented:
I've updated the PR to remove the remaining references to `multiprocessing`, instead using a second `ThreadPoolExecutor`. 

Also added in the dep on python-futures so hopefully the travis job will pass now.

TurboTurtle commented on sos/sosreport.py:
```diff
@@ -665,6 +678,10 @@ def _parse_args(self, args):
+        parser.add_option("-t", "--threads", action="store", dest="threads",
+                          help="specify number of concurrent plugins to run"
+                               " (default=4)",
+                          default=4)
```
Why yes, yes we do.

bmr-cymru commented:
Merged in commit 04df94418071b48a15aa80636dd34243ed374d2c
